File: solutions/go/need-for-speed/1/need_for_speed.go
package speed

// Car defines a remote controlled car with its specifications and state.
type Car struct {
	battery      int
	batteryDrain int
	speed        int
	distance     int
}

// NewCar creates a new remote controlled car with full battery and given specifications.
func NewCar(speed, batteryDrain int) Car {
	return Car{
		speed:        speed,
		batteryDrain: batteryDrain,
		battery:      100,
		distance:     0,
	}
}

// Track defines a race track with its distance.
type Track struct {
	distance int
}

// NewTrack creates a new track.
func NewTrack(distance int) Track {
	return Track{
		distance: distance,
	}
}

// Drive drives the car one time. If there is not enough battery to drive one more time,
// the car will not move.
func Drive(car Car) Car {
	// Check if there's enough battery to drive
	if car.battery >= car.batteryDrain {
		car.distance += car.speed
		car.battery -= car.batteryDrain
	}
	return car
}

// CanFinish checks if a car is able to finish a certain track.
func CanFinish(car Car, track Track) bool {
	// Calculate how many times the car can drive before battery is drained
	maxDrives := car.battery / car.batteryDrain
	
	// Calculate the maximum distance the car can cover
	maxDistance := maxDrives * car.speed
	
	// Check if the car can cover the track distance
	return maxDistance >= track.distance
}

File: solutions/r/need-for-speed/1/need-for-speed.R
new_car <- function(speed, battery_drain) {
  list(
    speed = speed,
    battery_drain = battery_drain,
    battery = 100,
    distance_traveled = 0
  )
}

new_track <- function(track_length) {
  list(
    track_length = track_length
  )
}

battery_drained <- function(car) {
  # Battery is drained if there's not enough to drive one more time
  car$battery < car$battery_drain
}

drive <- function(car) {
  # If the battery is drained, the car doesn't move
  if (battery_drained(car)) {
    return(car)
  }
  
  # Drive the car: increase distance, decrease battery
  car$distance_traveled <- car$distance_traveled + car$speed
  car$battery <- car$battery - car$battery_drain
  
  return(car)
}

can_finish <- function(car, track) {
  # Calculate how many times the car can drive before battery is drained
  max_drives <- car$battery %/% car$battery_drain
  
  # Calculate the maximum distance the car can cover
  max_distance <- max_drives * car$speed
  
  # Check if the car can cover the track distance
  return(max_distance >= track$track_length)
}

store_track <- function(car, track, name) {
  # Reset the car to full battery and zero distance
  car$battery <- 100
  car$distance_traveled <- 0
  
  # Determine if the car can finish the track
  can_complete <- can_finish(car, track)
  
  # Create the track sub-list
  track_info <- list(
    track_length = track$track_length,
    complete = can_complete
  )
  
  # Add the track info to the car using the track name
  car[[name]] <- track_info
  
  return(car)
}